dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/security/PublicKeysController.java

```diff
@@ -0,0 +1,124 @@
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.controller.security;
+
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+
+import java.security.interfaces.RSAPublicKey;
+import java.util.Map;
+
+import lombok.RequiredArgsConstructor;
+
+import org.hisp.dhis.common.DhisApiVersion;
+import org.hisp.dhis.security.oidc.DhisOidcClientRegistration;
+import org.hisp.dhis.security.oidc.DhisOidcProviderRepository;
+import org.hisp.dhis.webapi.mvc.annotation.ApiVersion;
+import org.springframework.security.oauth2.jose.jws.JwsAlgorithm;
+import org.springframework.security.oauth2.jose.jws.MacAlgorithm;
+import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jose.jwk.KeyType;
+import com.nimbusds.jose.jwk.KeyUse;
+import com.nimbusds.jose.jwk.RSAKey;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Controller
+@RequestMapping( "/publicKeys" )
+@RequiredArgsConstructor
+@ApiVersion( { DhisApiVersion.DEFAULT, DhisApiVersion.ALL } )
+public class PublicKeysController
+{
+
+    private final DhisOidcProviderRepository clientRegistrationRepository;
+
+    @GetMapping( value = "/{clientId}/jwks.json", produces = APPLICATION_JSON_VALUE )
+    public @ResponseBody Map<String, Object> getKeys( @PathVariable( "clientId" ) String clientId )
+    {
+        DhisOidcClientRegistration dhisOidcClientRegistration = clientRegistrationRepository
+            .getDhisOidcClientRegistration( clientId );
+
+        JWK jwk = dhisOidcClientRegistration.getJwk();
+        JwsAlgorithm jwsAlgorithm = resolveAlgorithm( jwk );
+        if ( jwsAlgorithm == null )
+        {
+            throw new IllegalArgumentException( "Unsupported algorithm: " + jwk.getAlgorithm() );
```

Comment:
Please try to think in terms of (and use) status code exceptions in controllers. Is this a conflict?